designer/server/src/main/scala/pl/touk/nussknacker/ui/process/processingtypedata/ProcessingTypeDataReload.scala

```diff
@@ -29,46 +16,31 @@ trait Initialization {
  * Another thing that needs careful consideration is handling exception during ProcessingTypeData creation/closing - probably during
  * close we want to catch exception and try to proceed, but during creation it can be a bit tricky...
  */
-class BasicProcessingTypeDataReload(
-    loadMethod: () => ProcessingTypeDataProvider[ProcessingTypeData, CombinedProcessingTypeData]
-) extends ProcessingTypeDataReload
+class ProcessingTypeDataReload(
+    loadMethod: () => ProcessingTypeDataState[ProcessingTypeData, CombinedProcessingTypeData]
+) extends ProcessingTypeDataProvider[ProcessingTypeData, CombinedProcessingTypeData]
     with LazyLogging {
 
-  @volatile private var current: ProcessingTypeDataProvider[ProcessingTypeData, CombinedProcessingTypeData] =
-    loadMethod()
-
-  override def reloadAll(): Unit = synchronized {
-    logger.info("Closing old models")
-    current.all(NussknackerInternalUser.instance).values.foreach(_.close())
-    logger.info("Reloading scenario type data")
-    current = loadMethod()
-    logger.info("Scenario type data reloading finished")
-  }
-
-}
-
-object BasicProcessingTypeDataReload {
-
-  def wrapWithReloader(loadMethod: () => ProcessingTypeDataProvider[ProcessingTypeData, CombinedProcessingTypeData]): (
-      ProcessingTypeDataProvider[ProcessingTypeData, CombinedProcessingTypeData],
-      ProcessingTypeDataReload with Initialization
-  ) = {
-    // must be lazy to avoid problems with dependency injection cycle - see NusskanckerDefaultAppRouter.create
-    lazy val reloader = new BasicProcessingTypeDataReload(loadMethod)
-    val provider = new ProcessingTypeDataProvider[ProcessingTypeData, CombinedProcessingTypeData] {
-      override def forType(processingType: ProcessingType)(implicit user: LoggedUser): Option[ProcessingTypeData] =
-        reloader.current.forType(processingType)
-
-      override def all(implicit user: LoggedUser): Map[ProcessingType, ProcessingTypeData] = reloader.current.all
-
-      override def combined: CombinedProcessingTypeData = reloader.current.combined
-    }
-    val lazyInitializedReloader = new ProcessingTypeDataReload with Initialization {
-      override def init(): Unit = reloader
-
-      override def reloadAll(): Unit = reloader.reloadAll()
-    }
-    (provider, lazyInitializedReloader)
+  @volatile private[processingtypedata] var state
```

Comment:
maybe we can use AtomicReference instead of volatile?

Reply:
`synchronized` blocks queued threads, so `AtomicReference` is better in this aspect. But at the same time, there are no effects in the `stateValue`, so the blocking is good enough.

Reply:
Yeah but on the other `AtomicReference`'s related operations can be processed many times so we should be sure that our transformation has no side effects. In the reload case it unfortunately has side effects.

Reply:
oh, I missed the `loadMethod` and its side-effect. So, synchronized FTW here ;)